### ISSUE TITLE
masked_conv2d BUG: the output seems to be wrong when masked shape is not matched with the output shape

### ISSUE THREAD

**Issue by ckirchhoff2021** [open-mmlab/mmcv#2282]
Hi, I am analyzing the masked_conv op, and i got some unexpected output when i feed some case to the op.The code is like below.
 
![image](https://user-images.githubusercontent.com/2441530/192090302-ef48c3f9-7920-4333-8967-31565ca307fb.png)

In my opinion, the mask is elementwised combined to the kernel ,and then compute the conv2d, thus the op is not accomplished like this.  It seems to be that the conv2d is calculated firstly then the mask is dotted to the output. the output fits my thoughts when the h and w of the output is same with the mask, like this
feature
![image](https://user-images.githubusercontent.com/2441530/192090643-94f7b5ab-da4d-43cc-ad66-e5ae4fcb3f78.png)
mask
![image](https://user-images.githubusercontent.com/2441530/192090653-f7fb38ef-9171-4680-8871-ad21f160dbbf.png)
masked_conv2d output
![image](https://user-images.githubusercontent.com/2441530/192090666-1b2fa3c4-11d6-4732-95cc-7b694da74248.png)
conv2d output
![image](https://user-images.githubusercontent.com/2441530/192090677-ff3cacbe-d13a-4991-8a7b-58c958adf36a.png)

when i modify the kernel size form 3 to 6, the w and h of output shuld be (7,7), while the mask is still (10,10), the confused output comes. 

features:
![image](https://user-images.githubusercontent.com/2441530/192091341-33ee1e82-690a-4ca0-85fb-aa7623ce3714.png)
kernels:
![image](https://user-images.githubusercontent.com/2441530/192091520-9f839bca-d4ad-4292-9075-a110edf68fed.png)
masked_conv2d output
![image](https://user-images.githubusercontent.com/2441530/192091531-d24688b4-c9f6-469f-ac31-6df419b5c38a.png)
conv2d output
![image](https://user-images.githubusercontent.com/2441530/192091538-b37edc6c-5aeb-4234-835d-177f02813704.png)
where the 550,420,285 at the masked_conv2d output row2 come? The values don't appear in the conv2d output ? can you explain your compute logic to me ? why is this?

**Comment by HAOCHENYE** [comment]
Hi, thanks for your feedback, can you provide the full script to reproduce the result you mentioned above?

**Comment by ckirchhoff2021** [comment]
import torch
import numpy as np
import torch.nn as nn
import torch.nn.functional as F
from mmcv.ops import masked_conv2d



def masked_conv_test():
    x = torch.arange(0,100).reshape([1,1,10,10]).cuda()
    print(x)
    x = x.float().cuda()
    
    mask_x = np.random.randint(0,2,100).reshape(1,10,10)
    print(mask_x)

    weight = torch.ones([1,1,6,6]).cuda()
    bias = torch.zeros([1]).cuda()
    # mask = torch.ones([1,10,10]).cuda()
    mask = torch.tensor(mask_x).cuda()
    
    padding = 1
    stride = 1

    y = masked_conv2d(x, mask, weight, bias, padding, stride)
    print(y.shape)
    print(y)

    y2 = F.conv2d(x, weight, bias, stride, padding)
    print(y2)
    print(y2.shape)

    # y3 = y2 * mask
    # print(y3)
    # print(y3.shape)


if __name__ == '__main__':
    masked_conv_test()


The code is like above, and I've made a photo in the issue description. You can run it by yourself, waiting for your solution.

**Comment by HAOCHENYE** [comment]
Sorry for my late reply, I actually spend some time figuring out why the behavior of `masked_conv2d` is so weird,  and now I have understood it :rofl:.

1. The masked convolution is not the implementation of paper https://paperswithcode.com/method/masked-convolution, but component of the paper https://arxiv.org/abs/1901.03278. So its mask it not for convolution kernal, but the output feature map.
2. It seems if the mask is for output feature map, why we directly multiply the output feature map by mask, but choosing the implement a cuda kernal? The answer is that the cuda kernel is more effective. For example, If we calculate the full feature map and multiply it with an mask, it means **10x10**(under your code above) times sliding window convolution is performed. However, only those positions where `mask[i, j] == 1` need to be convolved, therefore the cuda kernel will only compute the necessary result which will be more effective.
3. The cuda kernel actually has some implicit limitation, for example, the `input size`, `output size` and the `mask size` must be equal to the each other, otherwise the result will be unexpected. The process is a bit complicated, and the core reason is the index mismatch in cuda kernel. 

I think maybe we should assert the shape in python function, which could avoid such misleading behavior :tired_face:. Would mind helping us to fix it  :hand_over_mouth:
